zerver/models.py

```diff
@@ -152,6 +164,29 @@ class Realm(ModelReprMixin, models.Model):
     icon_version = models.PositiveSmallIntegerField(default=1)  # type: int
 
     DEFAULT_NOTIFICATION_STREAM_NAME = u'announce'
+    profile_fields = JSONStrField(default=list)  # type: List[Dict[str, Union[int, Text]]]
+
+    def get_profile_field_index(self, field_id):
+        # type: (int) -> int
+        """
+        `profile_fields` is a list of dictionaries; this function returns
+        the position (index) of a profile field in this list. Each element
+        has the following format and to get the postion we search for the
+        element which has the matching `field_id`. `field_id` is the primary
+        key of the model `CustomProfileField`.
+
+        .. code:: python
+            {
+                'id': <field_id>,
+                'name': <field name>,
+                'type': <field type>,
+            }
+        """
+        for i, field in enumerate(self.profile_fields):
+            if field['id'] == field_id:
+                return i
+
+        raise CustomProfileField.DoesNotExist()
```

Comment:
Comparing the code for the different approaches, I think the CustomProfileField and CustomProfileFieldValue models are a cleaner interface than the JSON thing.  It seems like you kinda have both implementations in here; let's get rid of the JSON field one and just keep the CustomProfileField/CustomProfileFieldValue models.